File: projeto/frontend/site/vox/src/app/services/auth.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable, map, tap } from 'rxjs';

export interface LoginResponse {
  token: string;
}

export interface UserProfile {
  id: number;
  email: string;
  name: string;
  fullname?: string;
  role: string;
  municipalityId: number;
  phone?: string;
  cpf?: string;
  birthDate?: string;
  profilePhotoUrl?: string | null;
}

export type UserRole = 'ADMINISTRATOR' | 'MODERATOR' | 'CITIZEN' | 'COUNCILOR';

export interface LogEntry {
  id: number;
  userId: number | null;
  userRole: string | null;
  municipalityId: number | null;
  httpMethod: string;
  path: string;
  queryString: string | null;
  statusCode: number;
  success: boolean;
  durationMs: number;
  ipAddress: string | null;
  userAgent: string | null;
  errorMessage: string | null;
  createdAt: string;
}

export interface LogPage {
  content: LogEntry[];
  page: number;
  size: number;
  totalElements: number;
}

export interface LogFilters {
  page: number;
  size: number;
  userId?: number;
  method?: string;
  from?: string;
  to?: string;
}

export interface RegisterPayload {
  name: string;
  email: string;
  cpf: string;
  phone?: string;
  password: string;
  birthDate?: string;
  municipalityId: number;
  acceptedTerms?: boolean;
  acceptedPrivacyPolicy?: boolean;
  file?: File | null;
}

export interface CreateUserPayload {
  name: string;
  email: string;
  cpf: string;
  phone?: string;
  password: string;
  birthDate?: string;
  role: UserRole | string;
  municipalityId: number;
  acceptedTerms?: boolean;
  acceptedPrivacyPolicy?: boolean;
}

@Injectable({
  providedIn: 'root'
})
export class AuthService {
  private readonly API_URL = 'http://localhost:8080';
  private readonly TOKEN_KEY = 'token';
  private readonly USER_ID_KEY = 'userId';
  private readonly MUNICIPALITY_ID_KEY = 'municipalityId';

  constructor(private http: HttpClient) {}

  login(email: string, password: string): Observable<LoginResponse> {
    return this.http.post<LoginResponse>(`${this.API_URL}/authenticate`, { email, password }).pipe(
      tap(response => {
        localStorage.setItem(this.TOKEN_KEY, response.token);
      })
    );
  }

  /**
   * Cadastro público de novo cidadão. O papel é sempre CITIZEN (não é
   * escolhido pelo usuário). Retorna a resposta do POST /api/user.
   */
  register(payload: RegisterPayload): Observable<unknown> {
    const formData = new FormData();
    const fields = {
      name: payload.name,
      email: payload.email,
      cpf: payload.cpf,
      phone: payload.phone ?? '',
      password: payload.password,
      birthDate: payload.birthDate ?? null,
      role: 'CITIZEN',
      municipalityId: payload.municipalityId,
      acceptedTerms: payload.acceptedTerms ?? true,
      acceptedPrivacyPolicy: payload.acceptedPrivacyPolicy ?? true
    };
    Object.entries(fields).forEach(([key, value]) => formData.append(key, String(value)));
    if (payload.file) formData.append('file', payload.file, payload.file.name);
    return this.http.post(`${this.API_URL}/api/user`, formData);
  }

  fetchCurrentUser(): Observable<UserProfile> {
    return this.http.get<Record<string, unknown>>(`${this.API_URL}/api/auth/me`).pipe(
      map(data => ({
        ...data,
        profilePhotoUrl: this.normalizePhotoUrl(
          data['profilePhotoUrl'] ?? data['profilePhoto'] ?? data['photoUrl']
        )
      } as UserProfile)),
      tap(user => {
        if (user?.id) {
          localStorage.setItem(this.USER_ID_KEY, String(user.id));
        }
        if (user?.municipalityId) {
          localStorage.setItem(this.MUNICIPALITY_ID_KEY, String(user.municipalityId));
        }
      })
    );
  }

  private normalizePhotoUrl(value: unknown): string | null {
    if (typeof value !== 'string' || !value.trim()) return null;
    if (/^https?:\/\//i.test(value)) return value;
    return `${this.API_URL}/${value.replace(/^\/+/, '')}`;
  }

  private getEmailFromToken(): string | null {
    const token = this.getToken();
    if (!token) return null;
    try {
      const payload = token.split('.')[1];
      const decoded = JSON.parse(atob(payload.replace(/-/g, '+').replace(/_/g, '/')));
      return decoded.email ?? decoded.sub ?? null;
    } catch {
      return null;
    }
  }

  logout(): void {
    localStorage.removeItem(this.TOKEN_KEY);
    localStorage.removeItem(this.USER_ID_KEY);
    localStorage.removeItem(this.MUNICIPALITY_ID_KEY);
  }

  isLoggedIn(): boolean {
    return !!localStorage.getItem(this.TOKEN_KEY);
  }

  getToken(): string | null {
    return localStorage.getItem(this.TOKEN_KEY);
  }

  getUserRole(): UserRole | null {
    const token = this.getToken();
    if (!token) return null;

    try {
      const payload = token.split('.')[1];
      const decoded = JSON.parse(atob(payload.replace(/-/g, '+').replace(/_/g, '/')));
      return decoded.role ?? decoded.roles?.[0] ?? null;
    } catch {
      return null;
    }
  }

  getUserId(): number | null {
    const stored = localStorage.getItem(this.USER_ID_KEY);
    if (stored) return Number(stored);
    // fallback: ler do payload do JWT
    const token = this.getToken();
    if (!token) return null;
    try {
      const payload = token.split('.')[1];
      const decoded = JSON.parse(atob(payload.replace(/-/g, '+').replace(/_/g, '/')));
      const id = decoded.userId ?? decoded.user_id ?? decoded.sub_id ?? null;
      return id != null ? Number(id) : null;
    } catch {
      return null;
    }
  }

  getMunicipalityId(): number {
    const stored = localStorage.getItem(this.MUNICIPALITY_ID_KEY);
    if (stored) return Number(stored);
    // fallback: ler do payload do JWT
    const token = this.getToken();
    if (!token) return 1;
    try {
      const payload = token.split('.')[1];
      const decoded = JSON.parse(atob(payload.replace(/-/g, '+').replace(/_/g, '/')));
      return decoded.municipalityId ?? decoded.municipality_id ?? 1;
    } catch {
      return 1;
    }
  }

  updateProfile(id: number, data: Partial<UserProfile>, file?: File | null): Observable<void> {
    const formData = new FormData();
    Object.entries(data).forEach(([key, value]) => {
      if (value !== undefined && value !== null) formData.append(key, String(value));
    });
    if (file) formData.append('file', file, file.name);
    return this.http.put<void>(`${this.API_URL}/api/user/${id}`, formData);
  }

  updatePassword(oldPassword: string, newPassword: string): Observable<void> {
    const id = this.getUserId();
    return this.http.put<void>(`${this.API_URL}/api/user/update-password`, {
      ...(id != null ? { id } : {}),
      oldPassword,
      newPassword
    });
  }

  forgotPassword(email: string): Observable<void> {
    return this.http.post<void>(`${this.API_URL}/api/auth/forgot-password`, { email });
  }

  resetPassword(token: string, newPassword: string): Observable<void> {
    return this.http.post<void>(`${this.API_URL}/api/auth/reset-password`, { token, newPassword });
  }

  getCouncilors(): Observable<UserProfile[]> {
    return this.http.get<UserProfile[]>(`${this.API_URL}/api/users/councilors`);
  }

  getCouncilorById(id: number): Observable<UserProfile> {
    return this.http.get<UserProfile>(`${this.API_URL}/api/users/councilors/${id}`);
  }

  // ----- Administração de usuários -----

  getAllUsers(): Observable<UserProfile[]> {
    return this.http.get<UserProfile[]>(`${this.API_URL}/api/user`);
  }

  getUsersByRole(role: UserRole | string): Observable<UserProfile[]> {
    return this.http.get<UserProfile[]>(`${this.API_URL}/api/user/role/${role}`);
  }

  getUserById(id: number): Observable<UserProfile> {
    return this.http.get<UserProfile>(`${this.API_URL}/api/user/${id}`);
  }

  createUser(payload: CreateUserPayload): Observable<unknown> {
    const formData = new FormData();
    Object.entries(payload).forEach(([key, value]) => {
      if (value !== undefined && value !== null) {
        formData.append(key, String(value));
      }
    });
    return this.http.post(`${this.API_URL}/api/user`, formData);
  }

  updateUser(id: number, data: Partial<UserProfile>): Observable<void> {
    const formData = new FormData();
    Object.entries(data).forEach(([key, value]) => {
      if (value !== undefined && value !== null) {
        formData.append(key, String(value));
      }
    });
    return this.http.put<void>(`${this.API_URL}/api/user/${id}`, formData);
  }

  deleteUser(id: number): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/api/user/${id}`);
  }

  // ----- Auditoria / logs -----

  getLogs(filters: LogFilters): Observable<LogPage> {
    let params = new HttpParams()
      .set('page', filters.page)
      .set('size', filters.size);
    if (filters.userId != null) params = params.set('userId', filters.userId);
    if (filters.method) params = params.set('method', filters.method);
    if (filters.from) params = params.set('from', filters.from);
    if (filters.to) params = params.set('to', filters.to);
    return this.http.get<LogPage>(`${this.API_URL}/api/admin/logs`, { params });
  }
}

File: projeto/frontend/app/app_vox/lib/services/auth_service.dart
import 'dart:convert';
import 'dart:typed_data';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import 'api_client.dart';
import '../models/user_profile.dart';

class AuthService {
  static const String _baseUrl = ApiClient.baseUrl;
  static const String _tokenKey = 'token';
  static const String _userIdKey = 'userId';
  static const String _municipalityIdKey = 'municipalityId';

  Future<String> login(String email, String password) async {
    final response = await http.post(
      Uri.parse('$_baseUrl/authenticate'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({'email': email, 'password': password}),
    );

    if (response.statusCode == 200) {
      final data = jsonDecode(response.body) as Map<String, dynamic>;
      final token = data['token'] as String;
      final prefs = await SharedPreferences.getInstance();
      await prefs.setString(_tokenKey, token);
      return token;
    } else if (response.statusCode == 401) {
      throw AuthException('E-mail ou senha inválidos.');
    } else {
      throw AuthException('Erro ao conectar ao servidor. Tente novamente.');
    }
  }

  Future<UserProfile> fetchCurrentUser() async {
    final response = await http.get(
      Uri.parse('$_baseUrl/api/auth/me'),
      headers: await ApiClient.authHeaders(),
    );
    ApiClient.checkResponse(response);
    final user = UserProfile.fromJson(jsonDecode(response.body) as Map<String, dynamic>);
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_userIdKey, user.id.toString());
    await prefs.setString(_municipalityIdKey, user.municipalityId.toString());
    return user;
  }

  Future<UserProfile> updateProfile(int id, Map<String, dynamic> data) async {
    final response = await http.put(
      Uri.parse('$_baseUrl/api/user/$id'),
      headers: await ApiClient.authHeaders(),
      body: jsonEncode(data),
    );
    ApiClient.checkResponse(response);
    return UserProfile.fromJson(jsonDecode(response.body) as Map<String, dynamic>);
  }

  Future<void> updatePassword(String currentPassword, String newPassword) async {
    final response = await http.put(
      Uri.parse('$_baseUrl/api/user/update-password'),
      headers: await ApiClient.authHeaders(),
      body: jsonEncode({
        'currentPassword': currentPassword,
        'newPassword': newPassword,
      }),
    );
    ApiClient.checkResponse(response);
  }

  Future<void> forgotPassword(String email) async {
    final response = await http.post(
      Uri.parse('$_baseUrl/api/auth/forgot-password'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({'email': email}),
    );
    ApiClient.checkResponse(response);
  }

  Future<void> resetPassword(String token, String newPassword) async {
    final response = await http.post(
      Uri.parse('$_baseUrl/api/auth/reset-password'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({'token': token, 'newPassword': newPassword}),
    );
    ApiClient.checkResponse(response);
  }

  Future<List<UserProfile>> getCouncilors() async {
    final response = await http.get(
      Uri.parse('$_baseUrl/api/users/councilors'),
      headers: await ApiClient.authHeaders(),
    );
    ApiClient.checkResponse(response);
    final list = jsonDecode(response.body) as List;
    return list
        .map((e) => UserProfile.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<UserProfile> getCouncilorById(int id) async {
    final response = await http.get(
      Uri.parse('$_baseUrl/api/users/councilors/$id'),
      headers: await ApiClient.authHeaders(),
    );
    ApiClient.checkResponse(response);
    return UserProfile.fromJson(jsonDecode(response.body) as Map<String, dynamic>);
  }

  Future<void> logout() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(_tokenKey);
    await prefs.remove(_userIdKey);
    await prefs.remove(_municipalityIdKey);
  }

  Future<String?> getToken() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(_tokenKey);
  }

  Future<bool> isLoggedIn() async {
    final token = await getToken();
    return token != null && token.isNotEmpty;
  }

  /// Decodifica o payload do JWT e retorna os campos (sem verificar assinatura).
  Map<String, dynamic> decodeToken(String token) {
    try {
      final parts = token.split('.');
      if (parts.length != 3) return {};
      String payload = parts[1];
      // Normaliza base64url → base64
      payload = payload.replaceAll('-', '+').replaceAll('_', '/');
      switch (payload.length % 4) {
        case 2:
          payload += '==';
          break;
        case 3:
          payload += '=';
          break;
      }
      final decoded = utf8.decode(Uint8List.fromList(base64Decode(payload)));
      return jsonDecode(decoded) as Map<String, dynamic>;
    } catch (_) {
      return {};
    }
  }

  Future<int?> getUserId() async {
    final prefs = await SharedPreferences.getInstance();
    final stored = prefs.getString(_userIdKey);
    if (stored != null) return int.tryParse(stored);
    final token = await getToken();
    if (token == null) return null;
    final payload = decodeToken(token);
    final id = payload['userId'] ?? payload['user_id'] ?? payload['sub_id'];
    return id != null ? int.tryParse(id.toString()) : null;
  }

  Future<int> getMunicipalityId() async {
    final prefs = await SharedPreferences.getInstance();
    final stored = prefs.getString(_municipalityIdKey);
    if (stored != null) return int.tryParse(stored) ?? 1;
    final token = await getToken();
    if (token == null) return 1;
    final payload = decodeToken(token);
    final id = payload['municipalityId'] ?? payload['municipality_id'];
    return id != null ? int.tryParse(id.toString()) ?? 1 : 1;
  }
}

class AuthException implements Exception {
  final String message;
  AuthException(this.message);

  @override
  String toString() => message;
}
